File: Hito_4/Observer/EjemploDetallado/VehicleRouting/handlers/BaseHandler.py
from IHandler import IHandler


class BaseHandler(IHandler):
    def __init__(self):
        self._next_handler: IHandler | None = None
        self._subscribers: list = []

    def set_next(self, next_handler: IHandler):
        self._next_handler = next_handler

    def is_solution_valid(self, solution: list[int]) -> bool:
        if self._next_handler:
            return self._next_handler.is_solution_valid(solution)
        self.update_subscribers("All good.")
        return True

    def add_subscriber(self, subscriber):
        if self._next_handler:
            self._next_handler.add_subscriber(subscriber)
        self._subscribers.append(subscriber)

    def update_subscribers(self, description: str):
        for subscriber in self._subscribers:
            subscriber.update(description)

File: Hito_4/Observer/EjemploDetallado/VehicleRoutingOriginal/VehicleRoutingSolucionadoObserver/VehicleRoutingValidator/Handlers/BaseHandler.cs
namespace VehicleRoutingValidator.Handlers;

public class BaseHandler : IHandler
{
    private IHandler? _nextHandler;
    private readonly List<Subscriber> _subscribers = new();
    
    public void SetNext(IHandler nextHandler)
        => _nextHandler = nextHandler;

    public virtual bool IsSolutionValid(int[] solution)
    {
        if(_nextHandler != null)
            return _nextHandler.IsSolutionValid(solution);
        UpdateSubscribers("All good.");
        return true;
    }

    public void AddSubscriber(Subscriber subscriber)
    {
        if (_nextHandler != null) 
            _nextHandler.AddSubscriber(subscriber);
        _subscribers.Add(subscriber);
    }

    protected void UpdateSubscribers(string description)
    {
        foreach (var subscriber in _subscribers)
        {
            subscriber.Update(description);
        }
    }
    
}